File: dailyQuestion/2020-05/05-28/python/solution_stack.py

# 利用栈(stack) 进行求解 进行
def decodeString(s: str) -> str:
    def is_num(num):
        if num >= '0' and num <= '9':
            return True
        return False

    stack = []
    for i in range(len(s)):
        if s[i] == "]":
            s1 = ""
            num = ""
            while stack[-1] != "[":
                s1 = stack.pop() + s1
            stack.pop()
            while len(stack) != 0 and is_num(stack[-1]):
                num = stack.pop() + num
            s1 = s1 * int(num)
            stack.append(s1)
        else:
            stack.append(s[i])
    return "".join(stack)

if __name__ == "__main__" :
    s = "3[a]2[bc]"
    result = decodeString(s)
    print(result)

File: dailyQuestion/2020/2020-05/05-28/golang/solution_stack.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	s := "100[leetcode]"
	result := decodeString(s)
	fmt.Println(result)

}

func decodeString(s string) string {
	stack := []string{}
	for i := 0; i < len(s); i++ {
		if s[i] != ']' {
			stack = append(stack, string(s[i]))
		} else {
			tmp_s := ""
			for {
				stack_len := len(stack)
				if stack[stack_len-1] != "[" {
					tmp_s = stack[stack_len-1] + tmp_s
					stack = stack[:stack_len-1]
				} else {
					break
				}
			}
			stack = stack[:len(stack)-1]
			repeat := ""
			for {
				stack_len := len(stack)
				if stack_len > 0 && stack[stack_len-1] >= "0" && stack[stack_len-1] <= "9" {
					repeat = stack[stack_len-1] + repeat
					stack = stack[:stack_len-1]
				} else {
					break
				}
			}
			_o, _ := strconv.ParseInt(repeat, 10, 64)
			_repeat := int(_o)
			tmp_t := strings.Repeat(tmp_s, _repeat)
			stack = append(stack, tmp_t)
		}
	}
	return strings.Join(stack, "")
}


